oak/server/rust/oak_runtime/src/runtime/graph.rs

```diff
@@ -0,0 +1,271 @@
+use crate::runtime::{ChannelHalfDirection, NodeId, Runtime};
+use itertools::Itertools;
+use std::{fmt::Write, string::String};
+
+/// Trait that gives an identifier for a data structure that is suitable for
+/// use with Graphviz/Dot.
+pub trait DotIdentifier {
+    fn dot_id(&self) -> String;
+}
+
+/// Trait that returns the path at which the debug introspection server will
+/// show a page for a data structure.
+pub trait HtmlPath {
+    fn html_path(&self) -> String;
+}
+
+impl DotIdentifier for NodeId {
+    fn dot_id(&self) -> String {
+        format!("node{}", self.0)
+    }
+}
+
+impl HtmlPath for NodeId {
+    fn html_path(&self) -> String {
+        format!("/node/{}", self.0)
+    }
+}
+
+impl DotIdentifier for (NodeId, oak_abi::Handle) {
+    fn dot_id(&self) -> String {
+        format!("{}_{}", self.0.dot_id(), self.1)
+    }
+}
+
+impl HtmlPath for (NodeId, oak_abi::Handle) {
+    fn html_path(&self) -> String {
+        format!("{}/{}", self.0.html_path(), self.1)
+    }
+}
+
+pub trait DotGraph {
+    fn graph(&self) -> String;
+
```

Comment:
My personal preference would be to skip the blank lines in this `trait` definition, but up to you.

Reply:
Thanks. Fixed.